docs/proposal/20210823-gpu-support.md

```diff
@@ -0,0 +1,163 @@
+
+#### Story 3 - PCI passthrough for single node Customer
+
+Alex is an Engineer at a retail organization that requires a single GPU node. They use one node with GPU attached and want to keep things simple. Alex can simply add this GPU-connected machine to the cluster and that should do the job. While selecting nodes, Alex can use appropriate labels to run his AI/ML workload on this particular node. PCI passthrough will provide direct GPU support. The challenge that Alex can face is that using passthrough Alex wouldn't be able to migrate nodes.
+
+### Requirements
+
+#### Functional Requirements
+
+* FR1: CAPV MUST support vGPUs.
+* FR2: CAPV MUST support GPUDirect.
+* FR3: CAPV MUST support PCI Passthrough.
+
+#### Non-Functional Requirements
+
+* NFR1: Unit tests MUST exist for all 3 supports
+* NFR2: e2e tests MUST exist for 3 supports
+
+### Implementation Details/Notes/Constraints
+
+#### Proposed Changes
+
+```go
+// VirtualMachineCloneSpec is information used to clone a virtual machine.
+type VirtualMachineCloneSpec struct {
+  // GPUProfileName is the Profile Name of a virtual machine's VGPU, in string.
+  // Defaults to the eponymous property value in the template from which the
+  // virtual machine is cloned.
+  // +optional
+  GPUProfileName string `json:"GPUProfileName,omitempty"`
+  // GPUDeviceId is the device ID of a virtual machine's GPU, in integer.
+  // Defaults to the eponymous property value in the template from which the
+  // virtual machine is cloned.
+  // +optional
+  GPUDeviceId int32 `json:"GPUDeviceId,omitempty"`
+  // GPUVendorId is the vendor ID of a virtual machine's GPU, in integer.
+  // Defaults to the eponymous property value in the template from which the
+  // virtual machine is cloned.
+  // +optional
+  GPUVendorId int32 `json:"GPUVendorId,omitempty"`
+}
+```
+
+#### Controller Changes
+
+* If VGPU or PCI device is present, the CRD is fetched and used to identify the expected devices.
+* The controller initiallizes variables respective to VGPU i.e `GPUProfileName`.
+* Similarly, controller also initializes variables related to Direct devices(for PCI passthrough and GPUDirect) ie. `GPUDeviceId` and `GPUVendorId`.
+* Using the above variables, respective objects of backing info of type `vimTypes.VirtualPCIPassthroughVmiopBackingInfo` with `{Vgpu: <GPUProfileName>}` for VGPU's and `vimTypes.VirtualPCIPassthroughAllowedDevice` for GPUDirect should be created by the controller.
+* Controller will call a function called as createPCI device which will use the information in above step to create objects of type `vimTypes.VirtualPCIPassthrough`
```

Comment:
should the VirtualMachineClone spec contain an array of PCI devices?

Reply:
I have added the fields into a GPU struct and loaded it inside of VirtualMachineClone as a slice.